lib/Service/ContactsIntegration.php

```diff
@@ -118,4 +118,113 @@ private function getPhotoUri(string $raw) {
 			return null;
 		}
 	}
+
+	/**
+	 * Adds a new email to an existing Contact
+	 *
+	 * @param string $uid
+	 * @param string $mailAddr
+	 * @param string $type
+	 * @return array
+	 */
+	public function addEmailToContact(string $uid, string $mailAddr, string $type = 'HOME') {
+		if (!$this->contactsManager->isEnabled()) {
+			return null;
+		}
+
+		if ($uid == null || $mailAddr == null) {
+			return null;
+		}
+
+		$result = $this->contactsManager->search($uid, ['UID'], ['types' => true]);
+		if (count($result) !== 1) {
+			return null;
+		}
+
+		$match = $result[0];
+		$email = $match['EMAIL'];
+
+		if (!is_array($email)) {
+			if (!isset($email) || $email === '') {
+				$email = [];
+			} else {
+				$email = [$email];
+			}
+		} else {
+			if (count($email) === 1 && $email[0]['value'] === '') {
+				$email = [];
+			}
+		}
+		$email[] = [ 
+			'type' => $type, 
+			'value' => $mailAddr 
+		];
+		$match['EMAIL'] = $email;
```

Comment:
If you limit your search result to 1 by adding 'limit' => 1 to the search options, you can work without all the ifs.

Something like:

```
		$result = $this->contactsManager->search($uid, ['UID'], ['types' => true, 'limit' => 1]);
		$email = $result[0]['EMAIL'] ?? [];
		if( empty($email) ) {
			return null;
		}
```
should do

Reply:
@miaulalala First of all, thank you for your review! I can see why you dislike the ifs. Tbh, I think this is sub-optimal too.
To check for an acceptable output, I wrote a few test cases. Long story short: I tried my best to write something more straight forward. There are still a couple of ifs, but maybe its clearer now. I also hope that the test cases make sense in some way or another and the generated output is okay.
```
$result1 = [];
$result2 = [
			[
				'UID' => 3,
				'FN' => 'Johann Strauss II',
			]
		];
$result3 = [
			[
				'UID' => 3,
				'FN' => 'Johann Strauss II',
				'EMAIL' => [
				    'type' => 'HOME',
				    'value' => 'jonathan@frakes.com'
				    ]
			]
		];
$result4 = [
			[
				'UID' => 3,
				'FN' => 'Johann Strauss II',
				'EMAIL' => [[
				    'type' => 'HOME',
				    'value' => 'jonathan@frakes.com'
				    ]]
			]
		];
$result5 = [
			[
				'UID' => 3,
				'FN' => 'Johann Strauss II',
				'EMAIL' => []
			]
		];
$result6 = [
	[
		'UID' => 3,
		'FN' => 'Johann Strauss II',
		'EMAIL' => [[]]
	]
];
$result7 = [
	[
		'UID' => 3,
		'FN' => 'Johann Strauss II',
		'EMAIL' => [[
				    'type' => 'HOME',
				    'value' => ''
		    ]]
	]
];
$result8 = [
	[
		'UID' => 3,
		'FN' => 'Johann Strauss II',
		'EMAIL' => [
				    'type' => 'HOME',
				    'value' => ''
		    ]
	]
];


$tests = [$result1, $result2, $result3, $result4, $result5, $result6, $result7, $result8];

foreach ($tests as $result) {
    print("test case:\n");
    print_r($result);
    if (count($result) !== 1) {
        print("no match\n");
        continue;
    }
    
    $newEntry = [
        'type' => "HOME",
        'value' => "test@mail"
    ];
    
    $email = $result[0]['EMAIL'] ?? [];
    print("email entry\n");
    print_r($email);
    if(!empty($email) && !is_array($email[0])) {
        $email = [$email];
    }
    $email[] = $newEntry;
    print("after adding \n");
    print_r($email);
}
```

Reply:
@BroCanDo Thanks for your response! Are those test cases all based on data from the contacts search result or is this you being super thorough and testing very well?

Reply:
@miaulalala They are partially (most of them) based on observed search results. Additionally I added a couple of edges cases which seemed plausible to me. Since I did not look at the implementation for the search, I thought that I rather handle a couple more edge cases.

Reply:
Ok, that's fine,, thanks for being so thorough!